### PR TITLE
Added Twitter login keys

### PR DESCRIPTION
**Pull request by st970703**
# Solution to #24
- **There were no access keys for Twitter**
- API Key and Secret created under my personal email
[st970703@gmail.com](st970703@gmail.com)
- **Twitter sign-in only works on localhost:8080**

Twitter sign-in:

1. Click Sign in -> Sign in with Twitter
2. Select account
3. (new account) Registration/new account page, clicking sign in again will redirect to the homepage
4. The homepage now displays "You are logged in as user ..."

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -224,6 +224,9 @@ bin/
 .cache-main
```
Comment (pulkitkalra) [inline .gitignore]: Looks like the merge conflicts in this file haven't been resolved, the conflict symbols are still there.

**Comment by victorlian** [comment]
Nice and simple fix :). No problem with the code. Twitter sign in is working! 👍 

However, upon registration an email should be sent (as there is a message saying an email is sent), yet this happens:


I digged into it a little bit, and I suspect the reason is the email field not being set:




In the case of facebook, the email field is set and there seems to be no exception.
In the case of google, registration email has an issue but i think it is now fixed ( #37 )

I'm not 100% sure whether this problem should be solved by you or should create another issue, but I have the feeling it belongs to this issue. Check with the TA/Kelly if you believe another issue should be created :).

**Comment by TheGuardianWolf** [comment]
If you commit in the secrets then they're no longer secret, I'm suggesting investigating alternative means of providing this information, otherwise your twitter API access could be abused.

**Comment by st970703** [comment]
After reading [Request a User’s Email Address](https://developer.twitter.com/en/docs/accounts-and-users/manage-account-settings/api-reference/get-account-verify_credentials), I have enabled the email access option in my Twitter App Developer account.

image

image

In the Twitter authorisation page, Twitter already stated that it will not access the user's email address.
I don't think that should be part of this issue.

This issue has already been discovered and is more to do with Twitter's policy:
https://stackoverflow.com/questions/37327302/spring-social-twitter-fetched-profile-returns-null-email-app-whitelisted

**Comment by st970703** [comment]
Approved by my team.
@softeng-701

**Comment by st970703** [comment]
Hey,
Please review this pull request.

@kblincoe

**Comment by kblincoe** [comment]
See my comment on issue #103. I do not think this code should be merged until the issue raised by TheGuardianWolf is fixed, but I do not think it should be its own issue as it is part of this issue.

**Comment by TheGuardianWolf** [comment]
To be fair issue wise, the other keys that were added all have the same issue. All of them needs this fix.

**Comment by st970703** [comment]
@TheGuardianWolf
That's why I think this issue should be separated out.
I can look into how to hide the Twitter ID and Secret.
But the Facebook and Google ones will still be exposed and can be abused?

**Comment by Nateeo** [comment]
@TheGuardianWolf @kblincoe @st970703 

I propose using `dotenv` or some other environment variable solution - this being an open source project, it would make sense to only allow people to drop-in their own API keys.

**Comment by TheGuardianWolf** [comment]
Yes, we can add to #50  or #92 an option to enter in the keys as an extension to the batch file.

**Comment by kblincoe** [comment]
We definitely don't want to ask users to provide an API key to be able to sign-in via Twitter. Signing in via another service is meant to make the sign-in process easier, not more difficult. I don't see how a combined issue will solve all three of these issues. I don't want to be merging in secret keys into this project.

**Comment by Nateeo** [comment]
I think we meant that anyone cloning and or deploying the application would
need to use their own keys to enable their users to log in. Using an
environment solution, the keys can be added from an env file or bash script
not checked into source control.

**Comment by TheGuardianWolf** [comment]
@st970703 just as a sidenote, make sure you invalidate the Twitter keys on your personal account, git history still keeps the keys.

Need a prune if you're interested in removing it, but that would be overkill.

**Comment by victorlian** [comment]
Hey code looks ok, works fine according to instructions, just one nitpicking thing:
the code can still be executed without the .properties file => lead to FileNotFound exception upon startup. 
Maybe include the .properties file but have some logic detecting that no key is supplied? Really don't like seeing the exceptions upon startup.

File: src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java [hunk 2]
```diff
@@ -95,8 +100,36 @@ public void addConnectionFactories(ConnectionFactoryConfigurer connectionFactory
...
+
+            // load a properties file
+            prop.load(input);
+
+            // get the property value and print it out
+            twitterClientId = prop.getProperty("twitter-client-id");
+            twitterClientSecret = prop.getProperty("twitter-client-secret");
+        } catch (IOException ex) {
+            ex.printStackTrace();
+        }
```
Comment (victorlian) [inline src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java]: And actually FileInputStreams should be closed as a good practice?

**Comment by victorlian** [comment]
I also found a little bug? Not sure if it is a bug.
But when you attempt to cancel a twitter sign in, you will be redirected to a page which is unauthorized:

**Steps to reproduce**
0. Have a twitter account that is logged in. (so you would not be prompt to log on to twitter)
1. Choose twitter sign in in QualOpt
2. Click on cancel when this show up:

3. Click on return to qualOpt.

4. This happens:


**Please investigate into this :)** @st970703

File: src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java [hunk 3]
```diff
@@ -95,8 +100,36 @@ public void addConnectionFactories(ConnectionFactoryConfigurer connectionFactory
         }
 
         // Twitter configuration
-        String twitterClientId = environment.getProperty("spring.social.twitter.client-id");
-        String twitterClientSecret = environment.getProperty("spring.social.twitter.client-secret");
+        String twitter_config_file_path = "src\\main\\resources\\config\\.twitter_config.properties";
+        this.configure_Twitter(connectionFactoryConfigurer, twitter_config_file_path);
+
+        // jhipster-needle-add-social-connection-factory
```
Comment (pulkitkalra) [inline src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java]: is this comment still necessary here?

Comment (st970703) [reply src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java]: It should still be there, as it helps to separate code into subsections.
Same goes for '//Google configuration', '//Facebook configuration'.

File: src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java (hunk 2 above)
Comment (st970703) [reply src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java]: Hey,
The code has been updated to close the FileInputStream.

Comment (victorlian) [reply src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java]: Thanks, please investigate on the bug :)

**Comment by st970703** [comment]
For 'You have not signed in to QualOpt.', the URL is 'https://api.twitter.com/oauth/authorize'.
This seems to be more on the Twitter side but I will investigate.

File: src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java (hunk 2 above)
Comment (pulkitkalra) [reply src/main/java/org/project36/qualopt/config/social/SocialConfiguration.java]: and maybe we could also log the exception? the Logger is already present in this class.

**Comment by Nateeo** [comment]
@st970703 I realise you've already done work on this so I'm sorry for weighing in so late - do you think we should create a common method for accessing social connection keys? I.E just make a single file and method for accessing the keys.

A possibly easier alternative would be to use the Java method `System.getenv(String environmentVariable)` to access environment variables containing the secrets.

**Comment by st970703** [comment]
Hey @Nateeo,
Sure. We can work on that together.

**Comment by st970703** [comment]
Hey @victorlian,

For the issue that the 'return to qualOpt' button is not working, according to what I saw in Chrome Dev Tool, that button is set to direct to a denied URL.
Hence it makes sense to see 'Your request cannot be processed.'

image

Unfortunately, this happened on the Twitter side and I should not be changing that.

If you click 'Go to the QualOpt homepage' link, it will lead you back to the QualOpt homepage correctly

**Comment by victorlian** [comment]
@st970703 ok... if there's nothing you could do about it, then I guess that's fine. Not a biggie anyway.
I'll approve :) 👍

**Comment by Nateeo** [comment]
@kblincoe 

@st970703  and I decided to approach secrets with environment variables - this would require the deployment platform such as the one QualOpt currently uses (Heroku) to configure environment variables, which simple involves filling out key value pairs. This would not require any user to do anything additional. Will contact and discuss with #39 (Github) to ensure we use an identical and agreed on approach.

**Comment by kblincoe** [comment]
Great, thanks! Is it ready to be merged?

**Comment by st970703** [comment]
Hey, @kblincoe,
I just changed my approach to using environment variables for Twitter sign-in.
The Spring Social Framework now will work on the Twitter Spring Social environment variables set by the user.
There is a batch file, 'batch_script\Setup_Spring_Social_App_Keys.bat' for the user to set their Twitter Spring Social environment variables on Windows.
The document of this change has been updated in 'QualOpt_SE701\docs\Twitter_Config.md'.

My teammates have agreed on this change and it can be merged.

**Pull request merged**